File: game-server/src/main/java/work/fking/pangya/game/GameServer.java
package work.fking.pangya.game;

import io.netty.channel.Channel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import work.fking.pangya.common.server.ServerConfig;
import work.fking.pangya.game.net.ServerChannelInitializer;
import work.fking.pangya.game.player.Item;
import work.fking.pangya.game.player.Player;
import work.fking.pangya.networking.SimpleServer;

import java.io.IOException;
import java.net.InetAddress;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;

public class GameServer {

    private static final Logger LOGGER = LoggerFactory.getLogger(GameServer.class);

    private final ExecutorService executorService = Executors.newVirtualThreadPerTaskExecutor();
    private final AtomicInteger connectionIdSequence = new AtomicInteger();
    private final AtomicInteger playerCount = new AtomicInteger();
    private final Map<Integer, Player> players = new ConcurrentHashMap<>();
    private final ServerConfig serverConfig;
    private final SessionClient sessionClient;
    private final List<ServerChannel> serverChannels;

    public GameServer(ServerConfig serverConfig, SessionClient sessionClient, List<ServerChannel> serverChannels) {
        this.serverConfig = serverConfig;
        this.sessionClient = sessionClient;
        this.serverChannels = serverChannels;
    }

    public int playerCount() {
        return playerCount.get();
    }

    public List<ServerChannel> serverChannels() {
        return Collections.unmodifiableList(serverChannels);
    }

    public ServerChannel serverChannelById(int id) {
        for (var serverChannel : serverChannels) {
            if (serverChannel.id() == id) {
                return serverChannel;
            }
        }
        return null;
    }

    public SessionClient sessionClient() {
        return sessionClient;
    }

    public void submitTask(Runnable runnable) {
        executorService.execute(runnable);
    }

    public Player findPlayer(int connectionId) {
        return players.get(connectionId);
    }

    public Player registerPlayer(Channel channel, int uid, String nickname) {
        var player = new Player(channel, uid, connectionIdSequence.incrementAndGet(), nickname);

        var characterRoster = player.characterRoster();
        characterRoster.unlockCharacter(67108872);

        var caddieRoster = player.caddieRoster();
        caddieRoster.unlockCaddie(469762048);

        var inventory = player.inventory();

        inventory.add(Item.create(268435511)); // clubset
        inventory.add(Item.create(335544382, 100)); // comets
        inventory.add(Item.create(436207656, 100)); // papel shop coupons

        // lucia items
        inventory.add(Item.create(136331265));
        inventory.add(Item.create(136423465));
        inventory.add(Item.create(136456205));
        inventory.add(Item.create(136456192));

        var equipment = player.equipment();
        equipment.equipCharacter(characterRoster.findByIffId(67108872));
        equipment.equipCaddie(caddieRoster.findByIffId(469762048));
        equipment.equipClubSet(inventory.findByIffId(268435511));
        equipment.equipComet(inventory.findByIffId(335544382));

        players.put(player.connectionId(), player);
        playerCount.incrementAndGet();
        channel.closeFuture().addListener(v -> onPlayerDisconnect(player));
        return player;
    }

    private void onPlayerDisconnect(Player player) {
        var channel = player.currentChannel();

        if (channel != null) {
            channel.removePlayer(player);
        }
        players.remove(player.connectionId());
        playerCount.decrementAndGet();
        LOGGER.debug("{} disconnected", player.nickname());
    }

    public void start() throws IOException, InterruptedException {
        var channelInitializer = ServerChannelInitializer.create(this);

        var server = SimpleServer.builder()
                                 .channelInitializer(channelInitializer)
                                 .address(InetAddress.getByName(serverConfig.bindAddress()))
                                 .port(serverConfig.port())
                                 .build();
        server.start();
    }
}

File: game-server/src/main/kotlin/work/fking/pangya/game/GameServer.kt
package work.fking.pangya.game

import io.netty.bootstrap.ServerBootstrap
import io.netty.channel.Channel
import io.netty.channel.ChannelOption
import org.slf4j.LoggerFactory
import work.fking.pangya.game.net.ServerChannelInitializer
import work.fking.pangya.game.persistence.PersistenceContext
import work.fking.pangya.game.player.CaddieRoster
import work.fking.pangya.game.player.CardInventory
import work.fking.pangya.game.player.CharacterRoster
import work.fking.pangya.game.player.Equipment
import work.fking.pangya.game.player.Inventory
import work.fking.pangya.game.player.Player
import work.fking.pangya.game.player.PlayerAchievements
import work.fking.pangya.game.player.PlayerGroup
import work.fking.pangya.game.player.PlayerStatistics
import work.fking.pangya.game.player.PlayerWallet
import work.fking.pangya.networking.selectBestEventLoopAvailable
import work.fking.pangya.networking.selectBestServerChannelAvailable
import java.net.InetAddress
import java.util.concurrent.Callable
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.concurrent.atomic.AtomicInteger

private val LOGGER = LoggerFactory.getLogger(GameServer::class.java)

class GameServer(
    private val serverConfig: GameServerConfig,
    val persistenceCtx: PersistenceContext,
    val sessionClient: SessionClient,
    val serverChannels: List<ServerChannel> = serverConfig.serverChannels
) {
    private val executorService = Executors.newThreadPerTaskExecutor(
        Thread.ofVirtual()
            .uncaughtExceptionHandler { _, throwable -> LOGGER.error("Uncaught exception while running task", throwable) }
            .factory()
    )
    private val connectionIdSequence = AtomicInteger()
    val players = PlayerGroup()

    fun serverChannelById(id: Int): ServerChannel? = serverChannels.firstOrNull { it.id == id }

    fun <R> submitTask(callable: Callable<R>): Future<R> = executorService.submit(callable)

    fun runTask(runnable: Runnable) {
        LOGGER.debug("Running task {}", runnable.javaClass.simpleName)
        executorService.execute(runnable)
    }

    fun registerPlayer(
        channel: Channel,
        sessionInfo: SessionClient.SessionInfo,
        playerWallet: PlayerWallet,
        characterRoster: CharacterRoster,
        caddieRoster: CaddieRoster,
        inventory: Inventory,
        cardInventory: CardInventory,
        equipment: Equipment,
        statistics: PlayerStatistics,
        achievements: PlayerAchievements,
    ): Player {
        val player = Player(
            channel = channel,
            uid = sessionInfo.uid,
            connectionId = connectionIdSequence.getAndIncrement(),
            username = sessionInfo.username,
            nickname = sessionInfo.nickname,
            wallet = playerWallet,
            characterRoster = characterRoster,
            caddieRoster = caddieRoster,
            inventory = inventory,
            cardInventory = cardInventory,
            equipment = equipment,
            statistics = statistics,
            achievements = achievements,
        )

        players.add(player)
        channel.closeFuture().addListener { onPlayerDisconnect(player) }
        return player
    }

    private fun onPlayerDisconnect(player: Player) {
        player.currentRoom?.removePlayer(player)
        player.currentChannel?.removePlayer(player)
        players.remove(player)
        LOGGER.info("{} disconnected", player.nickname)
    }

    fun start() {
        val bootstrap = ServerBootstrap()
        val bossGroup = selectBestEventLoopAvailable()
        val workerGroup = selectBestEventLoopAvailable(2)

        bootstrap.group(bossGroup, workerGroup)
            .channel(selectBestServerChannelAvailable())
            .childOption(ChannelOption.TCP_NODELAY, true)
            .childHandler(ServerChannelInitializer(this))

        LOGGER.info("Binding to port {}...", serverConfig.port)
        try {
            val inetAddress = InetAddress.getByName(serverConfig.bindAddress)
            val channel = bootstrap.bind(inetAddress, serverConfig.port)
                .sync()
                .channel()
            LOGGER.info("Successfully bound to {}, server bootstrap completed!", channel.localAddress())
            channel.closeFuture().sync()
        } finally {
            bossGroup.shutdownGracefully()
            workerGroup.shutdownGracefully()
        }
    }
}
